akka-http-core/src/main/scala/akka/http/util/StreamUtils.scala

```diff
@@ -193,6 +196,125 @@ private[http] object StreamUtils {
+    extends PublisherSink[In](attributes, shape) {
+    override def create(materializer: ActorFlowMaterializerImpl, flowName: String): (Subscriber[In], Publisher[In]) = {
+      val results = super.create(materializer, flowName)
+      cell.set(results._2)
+      results
+    }
+    override protected def newInstance(shape: SinkShape[In]): SinkModule[In, Publisher[In]] =
+      new OneTimePublisherSink[In](attributes, shape, cell)
+
+    override def withAttributes(attr: OperationAttributes): Module =
+      new OneTimePublisherSink[In](attr, amendShape(attr), cell)
+  }
+  /** A copy of SubscriberSource that allows access to the subscriber through the cell but can only materialized once */
+  private class OneTimeSubscriberSource[Out](val attributes: OperationAttributes, shape: SourceShape[Out], cell: OneTimeWriteCell[Subscriber[Out]])
+    extends SourceModule[Out, Subscriber[Out]](shape) {
+
+    override def create(materializer: ActorFlowMaterializerImpl, flowName: String): (Publisher[Out], Subscriber[Out]) = {
+      val processor = new Processor[Out, Out] {
+        @volatile private var subscriber: Subscriber[_ >: Out] = null
+
+        override def subscribe(s: Subscriber[_ >: Out]): Unit = subscriber = s
+        override def onError(t: Throwable): Unit = subscriber.onError(t)
+        override def onSubscribe(s: Subscription): Unit = subscriber.onSubscribe(s)
+        override def onComplete(): Unit = subscriber.onComplete()
+        override def onNext(t: Out): Unit = subscriber.onNext(t)
+      }
+      cell.set(processor)
+
+      (processor, processor)
+    }
+
+    override protected def newInstance(shape: SourceShape[Out]): SourceModule[Out, Subscriber[Out]] =
+      new OneTimeSubscriberSource[Out](attributes, shape, cell)
+    override def withAttributes(attr: OperationAttributes): Module =
+      new OneTimeSubscriberSource[Out](attr, amendShape(attr), cell)
+  }
+
+  trait ReadableCell[+T] {
+    def value: T
+  }
+  /** A one time settable cell */
+  class OneTimeWriteCell[T <: AnyRef] extends ReadableCell[T] {
+    @volatile var _value: T = null.asInstanceOf[T]
+    def value: T = {
+      require(_value != null, "Value wasn't set yet")
+      _value
+    }
+
+    def set(value: T): Unit = {
+      require(_value == null, "Value can be only set once.") // one time usage only
+      _value = value
+    }
+  }
+  /** Creates a flow out of sink and source */
+  def sinkSourceFlow[T, U](sink: Sink[T, Any], source: Source[U, Any]): Flow[T, U, Unit] =
+    Flow.wrap {
+      FlowGraph.partial() { implicit b ⇒
+        val in = b.add(sink)
+        val out = b.add(source)
+
+        FlowShape[T, U](in, out)
+      }
+    }
+
+  /** A merge for two streams that just forwards all elements and closes the connection eagerly. */
+  class EagerCloseMerge2[T](name: String) extends FlexiMerge[T, FanInShape2[T, T, T]](new FanInShape2(name), OperationAttributes.name(name)) {
```

Comment:
It is not yet clear to me why this is needed (and if it is then that would be a reason to just include this functionality as a config flag in the normal `Merge`).

From my understanding the WS stream can just be closed properly when the HTTP stream closes (e.g. by attaching an empty Publisher) and the HTTP handler stream can be closed when doing the WS upgrade. If these are true, then this merge variant would be unnecessary. The reason I’m arguing this point is because it somehow feels a little unclean to me, but I cannot quite put my finger onto the stain.